### ISSUE TITLE
The amount of the order does not match the total amount from the order lines

### ISSUE THREAD

**Issue by cusati-michael** [mollie/Shopware#271]
## Steps to reproduce

1. Create a product in a Shopware backend multiple customer groups and with specific prices per customer group.

Under Basic Settings -> Customer Groups -> (select a customer group)

<img src="https://i.ibb.co/yQN59c0/Screen-Shot-2021-09-29-at-15-55-46.png" width="500px">

2. Try to order and pay for the product as a customer from within the "discounted" customer group (in this case a -36% discount actually means a +36% surcharge), and Mollie will block the payment with the error:

```
Error executing API call (422: Unprocessable Entity): The amount of the order does not match the total amount from the order lines. Expected order amount to be CHF 30.59 but got CHF 34.39.
```

Later on in the error message, the line items used to calculate the expected value (CHF 30.59) are listed:

```
quantity: 2
unitPrice: CHF 5.30
totalAmount: CHF 10.60
shipping fee: CHF 19.99
```

The `unitPrice` should actually be CHF 7.20 (= 5.30 * 1.36). If I replace 5.30 with 7.20 the remaining calculations are all correct.

## Possible fix

Is there a simple way to make sure that Mollie uses the unitPrice according to the active Customer Group, the way that Shopware itself does? Thanks in advance for any thoughts!

**Comment by boxblinkracer** [comment]
Hi there

thank you for this
so the plugin does indeed use the final data that comes from Shopware, but there are some obstacles.

Mollie does always require gross prices....so with a shop configuration where taxes are charged, but in shopware only unit prices are in the data structure, then a manual conversion into gross prices need to be done....and also speaking of vertical or horizontal tax calculation, the Mollie API only allows 1 ...

long story short...theres an algorithm that converts into the formula required by Mollie.
problem is for example, that some article prices with more then 2 decimals could mean a slightly different end result...and theres even a hidden feature in the shopware backend to either round after tax calculation or not (strict rounding).
that has also be considered in that case.

and by now there are many unit tests also to verify this.

in your case, (i can see that due to the larger diff-amount) the answer might be very simple.
this is just a use case, nobody has thought of.
so I will add it to the official test cases, add unit tests and make sure that the customer group discount is also working (with whatever data shopware is coming up with)

thanks for this great hint

**Comment by cusati-michael** [comment]
@boxblinkracer Thanks so much for the quick response! I'll just keep an eye out for a patches/updates then. Please just let me know if there are any more details from my side that would be helpful for debugging this/ creating test cases, etc.

**Comment by boxblinkracer** [comment]
Hi
perfect
I just gave it a try, basically it worked for me, including discount. (btw you know that a negative discount makes it more expensive?)

so next thing i try is to have multiple prices for multiple customer groups.
there are indeed tests with a b2b customer group, so it might e a combination with discounts :D

do you have multiple groups with different prices? or at least 2?

**Comment by cusati-michael** [comment]
Yes, the "negative discount" is intended to be an additional expense. It sounds strange to me too, but it's the language used in Shopware 😅 

In our system there are roughly 20 customer groups with varying prices. Some of the customer groups are also designated as b2b, but I agree that the b2b designation doesn't seem to be the problem in and of itself.

I'm not sure if it matters, but the error occurred when testing payments in swiss francs (CHF). If it would be helpful, I could create a test customer group that simply uses EUR as a currency but still gets a -36% "discount", just to eliminate currency conversion as a variable here.

**Comment by boxblinkracer** [comment]
hi
perfect thank you
ill test it

the currency itself isnt a problem.
mollie does only do a cross check of the sent amounts, if they match.
this is completely independent from the currency

**Comment by boxblinkracer** [comment]
so
I have a customer with the b2b customer group in my shop.
then i have a custom price for a product for this customer group.
in addition to this, the customer group has a discount set.

the prices are all correct in the cart and on the confirm page.
then i click on "buy" and i get to the mollie sandbox page without an error and with the correct price

so i think we might need to do a zoom call.
are you in the shopware slack? could you reach out to me there?

**Comment by cusati-michael** [comment]
I just joined, but I'm not sure how to find you in there. I am Michael Ashton in the Slack group if you want to connect over there.

**Comment by boxblinkracer** [comment]
i there
i wrote you yesterday in slack, did you get it?
